api/server.go

```diff
@@ -13,17 +13,32 @@ import (
 	"io"
 	"net"
 	"net/http"
+	"os"
 	"strings"
 	"time"
 
 	"encoding/json"
 
+	gethLog "github.com/onflow/go-ethereum/log"
 	"github.com/onflow/go-ethereum/rpc"
 	"github.com/rs/cors"
 	"github.com/rs/zerolog"
 	"github.com/rs/zerolog/log"
 )
 
+var stderrHandler = gethLog.StreamHandler(
+	os.Stderr,
```

Comment:
is it ok to use stderr? would this logger be used for non-error logs as well?

Reply:
I have seen the following log line, for each served JSON-RPC request:

```json
{
  "conn": "127.0.0.1:45982",
  "duration": "29.641298ms",
  "lvl": "dbug",
  "msg": "Served eth_syncing",
  "reqid": "1",
  "t": "2024-05-14T11:10:00.562699232+03:00"
}
```
Do you want me to use another stream, such as `stdout`?

Reply:
could we use zerolog we use for all other outputs?

Reply:
I was able to create a custom log format type that resembles the output from zerolog in https://github.com/onflow/flow-evm-gateway/pull/243/commits/30d87aa44e35852f0554e2308abfaa524e3bc448

Reply:
I also made this format very specific to log only `error` log levels, since we want it only for crash reporting.

Reply:
The log line now looks like:
```json
{
  "component": "server",
  "level": "error",
  "message": "RPC method eth_syncing crashed: runtime error: index out of range [0] with length 0\ngoroutine 88 [running]:\ngithub.com/onflow/go-ethereum/rpc.(*callback).call.func1()\n\t/home/mpeter/go/pkg/mod/github.com/onflow/go-ethereum@v1.13.4/rpc/service.go:199 +0x89\npanic({0x196e780, 0xc0004aa4e0})\n\t/usr/local/go/src/runtime/panic.go:884 +0x213\ngithub.com/onflow/flow-evm-gateway/api.(*BlockChainAPI).Syncing(0xc0005529c0)\n\t/home/mpeter/Dev/flow-evm-gateway/api/api.go:120 +0x97\nreflect.Value.call({0xc000a88410?, 0xc0003de8b8?, 0x75e86006bc40?}, {0x1a47c00, 0x4}, {0xc000a774a0, 0x1, 0x0?})\n\t/usr/local/go/src/reflect/value.go:586 +0xb07\nreflect.Value.Call({0xc000a88410?, 0xc0003de8b8?, 0x1?}, {0xc000a774a0?, 0x0?, 0xc000167c50?})\n\t/usr/local/go/src/reflect/value.go:370 +0xbc\ngithub.com/onflow/go-ethereum/rpc.(*callback).call(0xc000a8d0e0, {0x1e43c38?, 0xc0001133b0}, {0xc000243c23, 0xb}, {0x2b67ae8, 0x0, 0x49efd7?})\n\t/home/mpeter/go/pkg/mod/github.com/onflow/go-ethereum@v1.13.4/rpc/service.go:205 +0x3c5\ngithub.com/onflow/go-ethereum/rpc.(*handler).runMethod(0xc000243c30?, {0x1e43c38?, 0xc0001133b0?}, 0xc0001b5ea0, 0x0?, {0x2b67ae8?, 0x4287f0?, 0x1834820?})\n\t/home/mpeter/go/pkg/mod/github.com/onflow/go-ethereum@v1.13.4/rpc/handler.go:565 +0x45\ngithub.com/onflow/go-ethereum/rpc.(*handler).handleCall(0xc000566960, 0xc000a773e0, 0xc0001b5ea0)\n\t/home/mpeter/go/pkg/mod/github.com/onflow/go-ethereum@v1.13.4/rpc/handler.go:512 +0x239\ngithub.com/onflow/go-ethereum/rpc.(*handler).handleCallMsg(0xc000566960, 0xc000a77440?, 0xc0001b5ea0)\n\t/home/mpeter/go/pkg/mod/github.com/onflow/go-ethereum@v1.13.4/rpc/handler.go:470 +0x237\ngithub.com/onflow/go-ethereum/rpc.(*handler).handleNonBatchCall(0xc000566960, 0xc000a773e0, 0xc0001b5ea0)\n\t/home/mpeter/go/pkg/mod/github.com/onflow/go-ethereum@v1.13.4/rpc/handler.go:296 +0x1ae\ngithub.com/onflow/go-ethereum/rpc.(*handler).handleMsg.func1.1(0x1e43c38?)\n\t/home/mpeter/go/pkg/mod/github.com/onflow/go-ethereum@v1.13.4/rpc/handler.go:269 +0x27\ngithub.com/onflow/go-ethereum/rpc.(*handler).startCallProc.func1()\n\t/home/mpeter/go/pkg/mod/github.com/onflow/go-ethereum@v1.13.4/rpc/handler.go:387 +0xc5\ncreated by github.com/onflow/go-ethereum/rpc.(*handler).startCallProc\n\t/home/mpeter/go/pkg/mod/github.com/onflow/go-ethereum@v1.13.4/rpc/handler.go:383 +0x8d\n",
  "time": "2024-05-17T16:22:50+03Z"
}
```

Reply:
What if we do something like this: 

```go
	gethLog.Root().SetHandler(gethLog.FuncHandler(func(r *gethLog.Record) error {
		switch r.Lvl {
		case gethLog.LvlInfo:
			zerolog.Info().Msg(r.Msg)
		case gethLog.LvlError:
			zerolog.Err(errors.New("api error")).Str("stack", r.Call.String())
		}
	}))
```

Reply:
also I would move this to `NewHTTPServer` not in the init.

Reply:
Moved the log wiring to the `NewHTTPServer` function in https://github.com/onflow/flow-evm-gateway/pull/243/commits/19ef6a24b3dbac084dfed8be5a6e31095274cf0e.

Reply:
I tried the code snippet you shared, but I can't see any logging.

Reply:
I managed to get your code snippet to work, it was not working due to not calling `Msg()` in the end.
Updated in https://github.com/onflow/flow-evm-gateway/pull/243/commits/d41ad4d1f2c7a1bf96677d33c324ad7ba8555a9a
Very good suggestion :clap:
This required very minimal changes to work, instead of creating a custom `JSONFormat`.